File: scripts/generate-tooth-icons.py
#!/usr/bin/env python3

from PIL import Image, ImageDraw
import random
import os

# Plaque color palette - using lighter browns from provided palette
plaque_colors = [
    (255, 234, 209),  # #FFEAD1
    (255, 228, 200),  # #FFE4C8
    (255, 225, 193),  # #FFE1C1
    (253, 187, 182),  # #FDBB6 (actually #FDBBB6)
    (255, 214, 172),  # #FFD6AC
]

def generate_tooth_icons():
    # Load the happy-teeth.png as base
    base_path = 'assets/timer/icons/happy-teeth.png'
    base_image = Image.open(base_path).convert('RGBA')
    width, height = base_image.size

    # Track configurations (tracks 2-8)
    # Increased plaque count for better visibility
    track_configs = [
        (2, 20),  # Top Left Back (1:45) - most plaque
        (3, 17),  # Top Right Front (1:30)
        (4, 14),  # Top Right Back (1:15)
        (5, 11),  # Bottom Left Front (1:00)
        (6, 8),   # Bottom Left Back (0:45)
        (7, 5),   # Bottom Right Front (0:30)
        (8, 2),   # Bottom Right Back (0:15) - least plaque
    ]

    for track_num, plaque_count in track_configs:
        # Create a copy of the base image
        img = base_image.copy()
        pixels = img.load()

        # First: Make EVERYTHING white except the tooth outline
        # We'll keep only the outer edge pixels and make everything else white
        for y in range(height):
            for x in range(width):
                r, g, b, a = pixels[x, y]

                # If pixel is visible
                if a > 0:
                    # Keep only if it's a very light gray or white (tooth edge/outline)
                    # Otherwise make it pure white
                    is_tooth_edge = False

                    # Check if it's on the very edge of the image
                    if x == 0 or x == width-1 or y == 0 or y == height-1:
                        # Keep transparent edge pixels as they are
                        if a < 255:
                            continue

                    # If it's a very light gray (tooth outline), keep it
                    if r > 250 and g > 250 and b > 250:
                        # Already white or near-white, keep as is
                        continue
                    elif r == g and g == b and r > 200 and r < 250:
                        # Light gray outline, keep it
                        continue

                    # Everything else becomes white (removes all blue, red, black pixels)
                    pixels[x, y] = (255, 255, 255, a)

        # Second: Find actual tooth pixels (white areas that are not transparent)
        # Build a map of valid tooth positions
        tooth_pixels = []
        for y in range(2, height - 2):  # Avoid very edge pixels
            for x in range(2, width - 2):
                r, g, b, a = pixels[x, y]
                # If it's a white pixel with full opacity (part of tooth)
                if a == 255 and r > 250 and g > 250 and b > 250:
                    # Check that it's not on the very edge of the tooth
                    # by ensuring neighboring pixels are also tooth pixels
                    neighbors_valid = True
                    for dx in [-1, 0, 1]:
                        for dy in [-1, 0, 1]:
                            if dx == 0 and dy == 0:
                                continue
                            nx, ny = x + dx, y + dy
                            if 0 <= nx < width and 0 <= ny < height:
                                nr, ng, nb, na = pixels[nx, ny]
                                # If any neighbor is transparent, this is an edge pixel
                                if na < 255:
                                    neighbors_valid = False
                                    break
                        if not neighbors_valid:
                            break

                    if neighbors_valid:
                        tooth_pixels.append((x, y))

        # Third: Add plaque squares only on valid tooth pixels
        draw = ImageDraw.Draw(img)
        plaque_size = 1  # Single pixel for small image

        placed_plaque = 0
        attempts = 0
        max_attempts = 1000
        used_positions = []

        while placed_plaque < plaque_count and attempts < max_attempts and tooth_pixels:
            attempts += 1

            # Pick a random tooth pixel
            x, y = random.choice(tooth_pixels)

            # Check if position overlaps with existing plaque
            too_close = False
            for px, py in used_positions:
                if abs(px - x) < plaque_size + 1 and abs(py - y) < plaque_size + 1:
                    too_close = True
                    break

            if not too_close:
                # Draw plaque square using colors from the palette
                color = plaque_colors[placed_plaque % len(plaque_colors)]
                draw.rectangle([x, y, x + plaque_size, y + plaque_size], fill=color)

                used_positions.append((x, y))
                placed_plaque += 1

        # Save the icon
        output_path = f'assets/timer/icons/tooth-track-{track_num}.png'
        img.save(output_path, 'PNG')
        print(f'Generated tooth-track-{track_num}.png with {placed_plaque} plaque squares')

    print('All tooth icons generated successfully!')

if __name__ == '__main__':
    generate_tooth_icons()

File: scripts/generate-tooth-icons.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { createCanvas, loadImage } = require('canvas');

// Plaque color palette (browns from light to dark)
const plaqueColors = [
  '#FFF3E9',
  '#FFF2E4',
  '#FFF2DE',
  '#FDECDA',
  '#FDE8D5',
  '#FFEAD1',
  '#FFE4C8',
  '#FFE1C1',
  '#FDBB6',
  '#FFD6AC'
];

async function generateToothIcons() {
  // Load the happy-teeth.png as base
  const basePath = path.join(__dirname, '../assets/timer/icons/happy-teeth.png');
  const baseImage = await loadImage(basePath);

  const canvas = createCanvas(baseImage.width, baseImage.height);
  const ctx = canvas.getContext('2d');

  // Track configurations (tracks 2-8, index 1-7)
  const trackConfigs = [
    { trackNum: 2, plaqueCount: 28 }, // Top Left Back (1:45)
    { trackNum: 3, plaqueCount: 24 }, // Top Right Front (1:30)
    { trackNum: 4, plaqueCount: 20 }, // Top Right Back (1:15)
    { trackNum: 5, plaqueCount: 16 }, // Bottom Left Front (1:00)
    { trackNum: 6, plaqueCount: 12 }, // Bottom Left Back (0:45)
    { trackNum: 7, plaqueCount: 8 },  // Bottom Right Front (0:30)
    { trackNum: 8, plaqueCount: 2 }   // Bottom Right Back (0:15)
  ];

  for (const config of trackConfigs) {
    // Clear canvas
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    // Draw base image
    ctx.drawImage(baseImage, 0, 0);

    // Get image data to modify pixels
    const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
    const data = imageData.data;

    // First pass: Remove the happy face (make it white)
    // We'll look for non-white, non-transparent pixels in the face area
    // and replace them with white
    for (let y = 0; y < canvas.height; y++) {
      for (let x = 0; x < canvas.width; x++) {
        const idx = (y * canvas.width + x) * 4;

        // Check if this pixel is part of the face (darker pixels in center area)
        // Assuming face is in middle area of teeth
        const centerX = canvas.width / 2;
        const centerY = canvas.height / 2;
        const distFromCenter = Math.sqrt(Math.pow(x - centerX, 2) + Math.pow(y - centerY, 2));

        // If pixel is dark (likely face) and in center area, make it white
        if (distFromCenter < 30) { // Adjust radius as needed
          const r = data[idx];
          const g = data[idx + 1];
          const b = data[idx + 2];
          const alpha = data[idx + 3];

          // If it's a dark pixel (face), make it white
          if (alpha > 0 && (r < 200 || g < 200 || b < 200)) {
            data[idx] = 255;     // R
            data[idx + 1] = 255; // G
            data[idx + 2] = 255; // B
            // Keep alpha as is
          }
        }
      }
    }

    // Put modified image data back
    ctx.putImageData(imageData, 0, 0);

    // Second pass: Add plaque squares
    // We'll add them randomly on the tooth surface
    const plaqueSize = 3; // Size of each plaque square in pixels
    const teethBounds = {
      // Define approximate tooth area bounds
      minX: 10,
      maxX: canvas.width - 10,
      minY: 10,
      maxY: canvas.height - 10
    };

    // Generate random positions for plaque
    const plaquePositions = [];
    const attempts = 0;
    const maxAttempts = 1000;

    while (plaquePositions.length < config.plaqueCount && attempts < maxAttempts) {
      const x = Math.floor(Math.random() * (teethBounds.maxX - teethBounds.minX - plaqueSize) + teethBounds.minX);
      const y = Math.floor(Math.random() * (teethBounds.maxY - teethBounds.minY - plaqueSize) + teethBounds.minY);

      // Check if this position is on a white area (tooth)
      const pixelIdx = (y * canvas.width + x) * 4;
      const pixelData = ctx.getImageData(x, y, 1, 1).data;

      // If pixel is white or near-white (tooth surface)
      if (pixelData[0] > 240 && pixelData[1] > 240 && pixelData[2] > 240 && pixelData[3] > 0) {
        // Check for overlap with existing plaque
        let overlap = false;
        for (const pos of plaquePositions) {
          if (Math.abs(pos.x - x) < plaqueSize + 2 && Math.abs(pos.y - y) < plaqueSize + 2) {
            overlap = true;
            break;
          }
        }

        if (!overlap) {
          plaquePositions.push({ x, y });
        }
      }
    }

    // Draw plaque squares
    plaquePositions.forEach((pos, index) => {
      // Pick a color from the palette
      const colorIndex = index % plaqueColors.length;
      ctx.fillStyle = plaqueColors[colorIndex];

      // Draw a small square
      ctx.fillRect(pos.x, pos.y, plaqueSize, plaqueSize);

      // Add some variation with slightly different sizes
      if (Math.random() > 0.5) {
        ctx.fillRect(pos.x + 1, pos.y + 1, plaqueSize - 1, plaqueSize - 1);
      }
    });

    // Save the icon
    const outputPath = path.join(__dirname, `../assets/timer/icons/tooth-track-${config.trackNum}.png`);
    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(outputPath, buffer);
    console.log(`Generated tooth-track-${config.trackNum}.png with ${config.plaqueCount} plaque squares`);
  }

  console.log('All tooth icons generated successfully!');
}

// Run the generation
generateToothIcons().catch(console.error);